### ISSUE TITLE
cura: weird non-deterministic corruption of larger meshes

### ISSUE THREAD

**Issue by gebner** [NixOS/nixpkgs#59901]
## Issue description

When loading larger STL files, cura sometimes corrupts them in seemingly random ways (every time different, very rarely the import works successfully):

![mutilated_benchy1](https://user-images.githubusercontent.com/313929/56435629-72811c00-62d9-11e9-9b5d-4d365083d2ca.png)
![mutilated_benchy2](https://user-images.githubusercontent.com/313929/56435631-74e37600-62d9-11e9-8f00-e7cfe7afc7be.png)

This is not just an OpenGL rendering problem, it also affects the sliced output:

![sliced_benchy1](https://user-images.githubusercontent.com/313929/56435649-8593ec00-62d9-11e9-9a4f-5762e11d8e90.png)

For comparison, this is how it should look like:

<img width="614" alt="expected_benchy" src="https://user-images.githubusercontent.com/313929/56435660-93497180-62d9-11e9-8036-aaf0c207c28b.PNG">

 - I cannot reproduce this issue on either ubuntu (with the PPA version of cura) or windows.
 - At first I thought this might be a problem with our OpenBLAS version (e.g. like #50407).  However there are no problems on ubuntu, which uses the same numpy and openblas versions.
 - The larger the STL file, the more likely the problems.  For example I could not succesfully import a larger STL file with 3 million faces at all.  (The same file shows no issues in either ubuntu or windows.)
 - The issue also occurs with 3MF files instead of STL.
 - I can also reproduce the issue with cura 3.6 on the stable channel.

This is the relevant part of the cura log output:
```
2019-04-19 20:31:01,074 - INFO - [Thread-3] STLReader.STLReader.load_file [46]: Using NumPy-STL to load STL data.
/nix/store/wmy9m8lam3ja21khh2yfp3crh4lb6pml-python3.7-uranium-4.0.0/lib/python3.7/site-packages/UM/Mesh/MeshData.py:436: RuntimeWarning: invalid value encountered in true_divide
  n[:, 0] /= l
/nix/store/wmy9m8lam3ja21khh2yfp3crh4lb6pml-python3.7-uranium-4.0.0/lib/python3.7/site-packages/UM/Mesh/MeshData.py:437: RuntimeWarning: invalid value encountered in true_divide
  n[:, 1] /= l
/nix/store/wmy9m8lam3ja21khh2yfp3crh4lb6pml-python3.7-uranium-4.0.0/lib/python3.7/site-packages/UM/Mesh/MeshData.py:438: RuntimeWarning: invalid value encountered in true_divide
  n[:, 2] /= l
2019-04-19 20:31:01,505 - DEBUG - [Thread-3] UM.Mesh.MeshData.calculateNormalsFromVertices [444]: Calculating normals took 0.03203701972961426 seconds
2019-04-19 20:31:01,519 - DEBUG - [Thread-3] STLReader.STLReader._read [97]: Loaded a mesh with 677118 vertices
2019-04-19 20:31:02,156 - DEBUG - [Thread-3] UM.Mesh.MeshData.approximateConvexHull [401]: approximateConvexHull(target_count=1024) Calculating 3D convex hull took 0.635796308517456 seconds. 21245 input vertices. 21245 output vertices.
/nix/store/rxvw3i2kmnjc3yc41xg4xfc0k2sygr2c-python3.7-numpy-1.16.1/lib/python3.7/site-packages/numpy/linalg/linalg.py:2480: RuntimeWarning: overflow encountered in multiply
  s = (x.conj() * x).real
/nix/store/wmy9m8lam3ja21khh2yfp3crh4lb6pml-python3.7-uranium-4.0.0/lib/python3.7/site-packages/UM/Mesh/MeshData.py:333: RuntimeWarning: invalid value encountered in true_divide
  data[:, 0] /= lengths
/nix/store/wmy9m8lam3ja21khh2yfp3crh4lb6pml-python3.7-uranium-4.0.0/lib/python3.7/site-packages/UM/Mesh/MeshData.py:334: RuntimeWarning: invalid value encountered in true_divide
  data[:, 1] /= lengths
/nix/store/wmy9m8lam3ja21khh2yfp3crh4lb6pml-python3.7-uranium-4.0.0/lib/python3.7/site-packages/UM/Mesh/MeshData.py:335: RuntimeWarning: invalid value encountered in true_divide
  data[:, 2] /= lengths
2019-04-19 20:31:02,280 - DEBUG - [Thread-3] UM.FileHandler.ReadFileJob.run [69]: Loading file took 1.2 seconds
```

Compared to ubuntu, the only differences are the `RuntimeWarning` and `Thread-3` (on ubuntu it's `Thread-1`).  The number of imported faces is also the same.

cc @abbradar @orivej 

### Steps to reproduce

1. Download a [benchy](https://www.thingiverse.com/download:1223854).
2. `nix-shell -p cura --run cura`
3. File/Open
4. Look at the corrupted model (you might need to repeat, sometimes it works)

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.19, NixOS, 19.09pre174426.acbdaa569f4 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(gebner): `""`
 - channels(root): `"nixos-19.09pre174426.acbdaa569f4"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by gebner** [comment]
Apparently there is an issue with OpenBLAS after all.  The problem goes away if I disable OpenMP support in `openblas.nix`:
```nix
USE_OPENMP = false;
```

@ttuegel @markuskowa @dtzWill Any ideas what is going on here?  Can we just disable OpenMP in OpenBLAS?

**Comment by markuskowa** [comment]
@gebner can you try if setting the environment variable "OMP_NUM_THREADS=1" already fixes the problem? There is also another bug which is related to AVX512 instructions https://github.com/NixOS/nixpkgs/issues/59708

**Comment by gebner** [comment]
Yes, setting the environment variable `OMP_NUM_THREADS=1` also solves the issue.

I don't think I have AVX512 instructions though (tested on an i7-6700K and i7-7500U).

**Comment by markuskowa** [comment]
Could this be related to the numpy issue we had in 18.09 https://github.com/NixOS/nixpkgs/issues/50407?

**Comment by gebner** [comment]
I thought the issue in #50407 has been fixed since openblas 0.3.4; at least the numpy devs seem to believe so: https://github.com/numpy/numpy/issues/12394#issuecomment-444179486

But yeah, it looks really similar to the bug in #50407 (i.e. https://github.com/xianyi/OpenBLAS/issues/1844).

**Comment by markuskowa** [comment]
@gebner Do you know by any chance, which blas routine causes the error?

**Comment by gebner** [comment]
Not really.  The numpy warnings come from this code:
 - https://github.com/Ultimaker/Uranium/blob/a441ac775f4bf488c554d476ad46b7781d75402e/UM/Mesh/MeshData.py#L432-L441
 - https://github.com/Ultimaker/Uranium/blob/a441ac775f4bf488c554d476ad46b7781d75402e/UM/Mesh/MeshData.py#L318-L335

**Comment by markuskowa** [comment]
These routines use `norm`, `dot`, and `cross`. That looks very similar to the previous problem with numpy/openblas (the openblas error appeared in combination with `dot`). Can you please open an issue upstream with openblas? Maybe we can get an upstream patch for the problem.

**Comment by nh2** [comment]
@gebner Could you try whether the merged openblas update fixes this even when

```nix
  makeWrapperArgs = [
    # hacky workaround for https://github.com/NixOS/nixpkgs/issues/59901
    "--set OMP_NUM_THREADS 1"
  ];
```

is removed?

**Comment by nh2** [comment]
I've tried it for a couple slices and didn't notice corruption.

In particular, the 3DBenchy from the screenshot works for me multiple tries in a row.

**Comment by nh2** [comment]
PR to remove the workaround: #82820

**Comment by gebner** [comment]
Sorry for taking a bit of time, I didn't feel too well the last few days.  I can still reproduce the issue in your branch: https://github.com/NixOS/nixpkgs/pull/82820#issuecomment-600898409

**Comment by charles-dyfis-net** [comment]
I've seen this in the curaLulzbot package too, and will be interested in whatever solution is landed on.

**Comment by charles-dyfis-net** [comment]
Strike that. I haven't seen it *recently*, and if my quite-hazy memory serves, it went away when I dropped a dependency (causing cura's built-in implementation to be used for the purpose at-hand).

`numpy-stl`, maybe? Unfortunately, https://code.alephobjects.com/ is down right now, so I can't reach the bugtracker entry associated with the bug the change was intended to fix.

**Comment by gebner** [comment]
> I've seen this in the curaLulzbot package too, and will be interested in whatever solution is landed on.

I have completely missed this comment.  We now have a workaround in the [cura package](https://github.com/NixOS/nixpkgs/blob/37379c0215c2f9296db7189cef2bdc74c6af2138/pkgs/applications/misc/cura/default.nix#L33-L36) by setting the `OMP_NUM_THREADS` environment variable to 1:
```nix
makeWrapperArgs = [
  # hacky workaround for https://github.com/NixOS/nixpkgs/issues/59901
  "--set OMP_NUM_THREADS 1"
];
```

**Comment by gebner** [comment]
The bug is still there, and the workaround continues to work around it.
